### ISSUE TITLE
RuntimeHandles should implement IEquatable

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtime#62944]
### Background and motivation

`RuntimeTypeHandle` is often used as a key in a dictionary. See hits on github: https://grep.app/search?q=dictionary%3Cruntimetypehandle. The search results include high profile libraries such a CsWinRt.

The runtime handles already implement strongly typed `Equals(XXXHandle other)` methods, but without the interfaces, dictionaries are going to use the `Equals(object other)` implementations instead, needlessly boxing the argument, unless a custom comparer is provided (it isn't in the above search results...).

### API Proposal

```C#
namespace System
{
    public struct RuntimeTypeHandle : IEquatable<RuntimeTypeHandle>
    {
        // bool Equals(RuntimeTypeHandle) already exists
    }

    public struct RuntimeMethodHandle : IEquatable<RuntimeMethodHandle>
    {
        // bool Equals(RuntimeMehtodHandle) already exists
    }

    public struct RuntimeFieldHandle : IEquatable<RuntimeFieldHandle>
    {
        // bool Equals(RuntimeFieldHandle) already exists
    }
}
```

### API Usage

No boxing.

**Comment by terrajobst** [comment]
* Looks good as proposed
* We should do a pass and find all structs that provide a custom `Equals` method without implementing `IEquatable<RuntimeTypeHandle>`
* This should probably be an analyzer

```C#
namespace System
{
    public struct RuntimeTypeHandle : IEquatable<RuntimeTypeHandle>
    {
        // bool Equals(RuntimeTypeHandle) already exists
    }

    public struct RuntimeMethodHandle : IEquatable<RuntimeMethodHandle>
    {
        // bool Equals(RuntimeMethodHandle) already exists
    }

    public struct RuntimeFieldHandle : IEquatable<RuntimeFieldHandle>
    {
        // bool Equals(RuntimeFieldHandle) already exists
    }
}
```

**Comment by jkotas** [comment]
> This should probably be an analyzer

https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1066 ?

**Comment by stephentoub** [comment]
>> This should probably be an analyzer

> https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1066 ?

Yes,  we need to fix or suppress all of our occurrences and then turn it on:
https://github.com/dotnet/runtime/issues/56082#issuecomment-900661055

**Comment by MichalStrehovsky** [comment]
Fixed in #63358. IEquatable on all the other structs that have Equals should be a different issue, if we want such issue. Looks like one of those issues that are so big that we'll never get to them.

**Comment by stephentoub** [comment]
> IEquatable on all the other structs that have Equals should be a different issue, if we want such issue. Looks like one of those issues that are so big that we'll never get to them.

I opened https://github.com/dotnet/runtime/issues/63687 with the list for API review and https://github.com/dotnet/runtime/pull/63690 with an implementation.
